### PR TITLE
Add way to discriminate between compression standards in the cache

### LINKED ISSUE #9: Add compression to csv datasets
This would save lots of disk space for relatively little performance cost.
Also, Arrow can read gzip directly.

(De)compressing CSV files should be a conversion just like #49 adds for parquet. So it should be possible to have both the compressed and uncompressed variants of a CSV dataset in the cache.

### LINKED ISSUE #16: Add way to discriminate between parquet compression standards in the cache
When converting a dataset to parquet, the user can provide a compression standard that is passed to the parquet writer. But downloaded datasets in parquet format already have a certain compression. Currently the compression standard is not in the cache hierarchy naming structure, so it is not possible to distinguish a datasets wrt its parquet compression. We should consider adding this

### PR DESCRIPTION
**Pull request by joosthooz**
Fixes #9, fixes #16
This adds a directory hierarchy level, to denote what internal compression is used.
For downloaded parquet files, it first downloads to a directory `parquetcompression_unknown`, then detects the compression and moves it to the proper location in the cache.
- [x] Parquet
- [x] csv

### CODE REVIEW

File: tests/test_datalogistik.py [hunk 1]
```diff
@@ -131,7 +131,9 @@ def test_convert_dataset_csv_to_parquet():
     written_table = csv.read_csv(test_csv_file_path)
     print(written_table.schema)
     assert written_table == orig_table
-    converted_path = util.convert_dataset(dataset_info, None, "csv", "parquet", 0, 0)
+    converted_path = util.convert_dataset(
+        dataset_info, None, None, "csv", "parquet", 0, 0
+    )
```
Comment (austin3dickey) [inline tests/test_datalogistik.py:136]: I assume these tests cover a lot of this new code, correct? Is there value in adding another test that converts between different compression formats?

**Review by austin3dickey — COMMENTED** [review]
Looking great!

**Comment by joosthooz** [comment]
Now both parquet and csv have a cache directory hierarchy level for compression. Conversion between gzip and uncompressed work both ways, but I'm going to change that into re-downloading the file because it is much faster

**Comment by alistaire47** [comment]
> Conversion between gzip and uncompressed work both ways, but I'm going to change that into re-downloading the file because it is much faster

This is definitely not true for all datasets; some of the taxi variants can take >1h to download but gzip/decompress in minutes. And why wouldn't we [usually] keep the downloaded version around anyway, since that's the source of truth? Or keep a quick-to-read version around (probably parquet-snappy) so the primary cost is just resaving?

File: datalogistik/cli.py [hunk 2]
```diff
@@ -75,7 +75,8 @@ def parse_args():
         "-c",
         "--compression",
         type=str,
-        help="Internal compression (passed to parquet writer)",
+        help="Compression (for parquet: passed to parquet writer, "
+        "for csv: either None of gz)",
```
Comment (austin3dickey) [inline datalogistik/cli.py:79]:
```suggestion
        "for csv: either None or gz)",
```

File: datalogistik/cli.py [hunk 3]
```diff
@@ -165,11 +166,19 @@ def parse_args_and_get_dataset_info():
...
-            if opts.format != "parquet":
-                msg = "Compression is only supported for parquet format"
-                log.error(msg)
-                raise ValueError(msg)
+        # Defaults
+        if opts.compression is None:
+            if opts.format == "parquet":
+                opts.compression = "snappy"
+            if opts.format == "csv":
+                opts.compression = "gz"
```
Comment (austin3dickey) [inline datalogistik/cli.py:174]: Should we be using `tar.gz`? I think it's more standard for that to be the file extension, no?

File: tests/test_datalogistik.py (hunk 1 above)
Comment (austin3dickey) [reply tests/test_datalogistik.py:428]: It's hard for me to tell from this file, but did you cover compressing/decompressing csvs and parquets in these tests?

Also, perhaps there's value in adding compression to our integration tests?

**Review by austin3dickey — CHANGES_REQUESTED** [review]
Nice, a few small things. I also agree with Ed that we should probably prefer converting a local dataset instead of re-downloading it, even at the cost of speed, because fewer network requests is probably a good idea

**Comment by joosthooz** [comment]
I added some tests, but there's still a job to be done there; to be able to test all kinds of conversions between format, partitioning, and compression, we first need to generate a dataset with only datatypes that both parquet and csv support. I'll make a note of that

**Review by austin3dickey — COMMENTED** [review]
I see the unit tests are sometimes failing with
```
Differing items:
{'local-creation-date': '2022-09-05T10:40:14+0000'} != {'local-creation-date': '2022-09-05T10:40:13+0000'}
```
Will that be fixed in #46 or can you fix it here?

**Comment by joosthooz** [comment]
> Will that be fixed in https://github.com/conbench/datalogistik/pull/46 or can you fix it here?

Yes, it is fixed in #46

**Review by austin3dickey — APPROVED** [review]
LGTM

**Pull request merged**